homeassistant/components/onkyo/media_player.py

```diff
@@ -105,417 +185,361 @@
+        active_zones.append(OnkyoAVR(avr, name, sources, zone, max_volume))
+
+    @callback
+    def close_avr(_event):
+        avr.close()
+
+    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, close_avr)
+
+    async_add_entities(active_zones)
+
+
+class OnkyoAVR(MediaPlayerEntity):
+    """Entity reading values from Onkyo AVR protocol."""
+
+    def __init__(self, avr, name, sources, zone, max_volume):
+        """Initialize entity with transport."""
+        super().__init__()
+        self._avr = avr
+        self._name = f"{name} {zone if zone != 'main' else ''}"
+        self._zone = zone
+        self._volume = 0
+        self._supports_volume = False
+        self._muted = False
         self._max_volume = max_volume
-        self._receiver_max_volume = receiver_max_volume
-        self._current_source = None
+        self._powerstate = STATE_UNKNOWN
+        self._source = None
         self._source_list = list(sources.values())
         self._source_mapping = sources
         self._reverse_mapping = {value: key for key, value in sources.items()}
+        self._sound_mode = None
         self._attributes = {}
-        self._hdmi_out_supported = True
-
-    def command(self, command):
-        """Run an eiscp command and catch connection errors."""
-        try:
-            result = self._receiver.command(command)
-        except (ValueError, OSError, AttributeError, AssertionError):
-            if self._receiver.command_socket:
-                self._receiver.command_socket = None
-                _LOGGER.debug("Resetting connection to %s", self._name)
-            else:
-                _LOGGER.info("%s is disconnected. Attempting to reconnect", self._name)
-            return False
-        _LOGGER.debug("Result for %s: %s", command, result)
-        return result
-
-    def update(self):
-        """Get the latest state from the device."""
-        status = self.command("system-power query")
-
-        if not status:
-            return
-        if status[1] == "on":
-            self._pwstate = STATE_ON
-        else:
-            self._pwstate = STATE_OFF
-            return
-
-        volume_raw = self.command("volume query")
-        mute_raw = self.command("audio-muting query")
-        current_source_raw = self.command("input-selector query")
-        # If the following command is sent to a device with only one HDMI out,
-        # the display shows 'Not Available'.
-        # We avoid this by checking if HDMI out is supported
-        if self._hdmi_out_supported:
-            hdmi_out_raw = self.command("hdmi-output-selector query")
-        else:
-            hdmi_out_raw = []
-        preset_raw = self.command("preset query")
-        if not (volume_raw and mute_raw and current_source_raw):
+        self._supports_sound_mode = False
+        self._query_timer = None
+
+    async def async_added_to_hass(self):
+        """Entity has been added to hass."""
+        self.backfill_state()
+
```

Comment:
```suggestion

    async def async_will_remove_from_hass(self):
        """Cancel the query timer when the entity is removed."""
        if self._query_timer:
            self._query_timer()
            self._query_timer = None
```

Reply:
I'll add this in a new commit to add an extra new line after `self._query_timer = None`